mdolab/pyoptsparse
# Optview rewrite

## Purpose

This PR introduces a new version of OptView that uses an MVC design pattern and PtQt5, built on the pyOptSparse history file api.  This new version will fully replace the old version of OptView and add PyQt5 and matplotlib as dependencies if installed with the `pip install .[optview]` tag.  OpenMDAO case recorder files will no longer be supported in the new version.  Features include:
- Updated user interface.
- Command line interface for adding files on startup.
- Support for multiple history files.
- Support for plotting major and minor iterations.
- Ability to add multiple plots and configure them individually.
- Optimization problem metadata viewer with searchable options list.
- Support for multiple plotting tabs that can support up to three subplots each.
- Default niceplots formatting for matplotlib.
- Support for saving plots as pdf, pgf, jpeg, png, and eps using the matplotlib Qt5Agg backend.
- Plot configuration toolbar from the matplotlib Qt5Agg backend.
- All classes and functions are fully documented with docstrings and comments!

## Expected time until merged

This PR is quite large and introduces a lot of postprocessing functionality that did not previously exist.  I expect the PR to take minimum a week, but likely more.

## Type of change


- [ ] Breaking change (non-backwards-compatible fix or feature)

This change will not affect the main functionality of pyOptSparse, but will completely replace the old version of OptView.

## Testing


This PR introduces a working version of the new OptView, but still needs lots of test coverage.  I plan to add tests for each function in the future, but for now, to check out the functionality, do these steps:

1. Navigate to the postprocessing directory.
2. Run `python OptView.py` in a terminal.

## Checklist


- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
